docs/language/ql-handbook/types.rst

```diff
@@ -479,6 +480,48 @@ program, so it's helpful to extend a new type (namely ``TTaintType``)::
     class Tainted extends TaintType, TTaintedValue {
     }
     
+.. _type-unions:
+
+Type unions
+***********
+
+Type unions are user-defined types that are declared with the keyword ``class``.
+The syntax resembles :ref:`type aliases <type-aliases>`, but with two or more type expressions on the right-hand side.
+
+Type unions are used for creating restricted subsets of an existing :ref:`algebraic datatype <algebraic-datatypes>`, by explicitly
+selecting a subset of the branches of that datatype and binding them to a new type.
+Type unions of :ref:`database types <database-types>` are also supported.
+
+You can use a type union to give a name to a subset of the branches from an algebraic datatype.
+In some cases, using the type union over the whole algebraic datatype can avoid spurious
+:ref:`recursion <recursion>` in predicates.
+For example, the following construction is legal::
+
+    newtype InitialValueSource =
+        ExplicitInitialization(VarDecl v) { exists(v.getInitializer()) } or
+        ParameterPassing(Call c, int pos) { exists(c.getParameter(pos)) } or
+        UnknownInitialGarbage(VarDecl v) { not exists(DefiniteInitialization di | v = target(di)) }
+
+    class DefiniteInitialization = ParameterPassing or ExplicitInitialization;
+
+    VarDecl target(DefiniteInitialization di) {
+       di = ExplicitInitialization(result) or
+       exists(Call c, int pos | di = ParameterPassing(c, pos) and
+                                result = c.getCallee().getFormalArg(pos))
+    }
+
+However, a similar implementation that restricts ``InitialValueSource`` in a class extension is not valid.
+If we had implemented ``DefiniteInitialization`` as a class extension instead, it would trigger a type test for ``InitialValueSource``. This results in an illegal recursion ``DefiniteInitialization -> InitialValueSource -> UnknownInitialGarbage -> ¬DefiniteInitialization`` since ``UnknownInitialGarbage`` relies on ``DefiniteInitialization``::
```

Comment:
I'm still slightly worried about the distance between the hypothetical wording "if we had implemented" and the code block. Readers in a hurry could easily miss the former and take the example as an example of good QL (or even assume that it's what the union type definition _desugars to_). Can we at least add a `// THIS WON'T WORK` to the code block?

Reply:
thanks for the suggestion, I did that